Source/Diagnostics/ReducedDiags/Costs.cpp

```diff
@@ -0,0 +1,199 @@
+    }
+
+    // parallel reduce to IO proc and get data over all procs
+    amrex::Vector<amrex::Real>::iterator it_m_data = m_data.begin();
+    amrex::Real* addr_it_m_data = &(*it_m_data);
+    ParallelReduce::Sum(addr_it_m_data,
+                        m_data.size(),
+                        ParallelDescriptor::IOProcessorNumber(),
+                        ParallelContext::CommunicatorSub());
+
+    /* m_data now contains up-to-date values for:
+     *  [[cost, proc, lev, i_low, j_low, k_low] of box 0 at level 0,
+     *   [cost, proc, lev, i_low, j_low, k_low] of box 1 at level 0,
+     *   [cost, proc, lev, i_low, j_low, k_low] of box 2 at level 0,
+     *   ...
+     *   [cost, proc, lev, i_low, j_low, k_low] of box 0 at level 1,
+     *   [cost, proc, lev, i_low, j_low, k_low] of box 1 at level 1,
+     *   [cost, proc, lev, i_low, j_low, k_low] of box 2 at level 1,
+     *   ......] */
+
+}
+// end void Costs::ComputeDiags
+
+// write to file function for cost
+void Costs::WriteToFile (int step) const
+{
+    ReducedDiags::WriteToFile(step);
+
+    // get WarpX class object
+    auto& warpx = WarpX::GetInstance();
+
+    if (ParallelDescriptor::IOProcessor())
+    {
+        // final step is a special case, fill jagged array with NaN
+        if (step == (warpx.maxStep() - 1))
+        {
+            // open tmp file to copy data
+            std::string fileTmpName = m_path + m_rd_name + ".tmp." + m_extension;
+            std::ofstream ofs(fileTmpName, std::ofstream::out);
+
+            // write header row
+            // for each box on each level we saved 6 data fields: [cost, proc, lev, i_low, j_low, k_low])
+            ofs << "#";
+            ofs << "[1]step()";
+            ofs << m_sep;
+            ofs << "[2]time(s)";
+            ofs << m_sep;
+            for (int boxNumber=0; boxNumber<nBoxesMax; ++boxNumber)
+            {
+                ofs << "[" + std::to_string(3 + nDataFields*boxNumber) + "]";
+                ofs << "cost_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+                ofs << "[" + std::to_string(4 + nDataFields*boxNumber) + "]";
+                ofs << "proc_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+                ofs << "[" + std::to_string(5 + nDataFields*boxNumber) + "]";
+                ofs << "lev_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+                ofs << "[" + std::to_string(6 + nDataFields*boxNumber) + "]";
+                ofs << "i_low_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+                ofs << "[" + std::to_string(7 + nDataFields*boxNumber) + "]";
+                ofs << "j_low_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+                ofs << "[" + std::to_string(8 + nDataFields*boxNumber) + "]";
+                ofs << "k_low_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+            }
```

Comment:
```suggestion
            for (int boxNumber=0; boxNumber<nBoxesMax; ++boxNumber)
            {
                ofs << m_sep;
                ofs << "[" + std::to_string(3 + nDataFields*boxNumber) + "]";
                ofs << "cost_box_"+std::to_string(boxNumber)+"()";
                ofs << m_sep;
                ofs << "[" + std::to_string(4 + nDataFields*boxNumber) + "]";
                ofs << "proc_box_"+std::to_string(boxNumber)+"()";
                ofs << m_sep;
                ofs << "[" + std::to_string(5 + nDataFields*boxNumber) + "]";
                ofs << "lev_box_"+std::to_string(boxNumber)+"()";
                ofs << m_sep;
                ofs << "[" + std::to_string(6 + nDataFields*boxNumber) + "]";
                ofs << "i_low_box_"+std::to_string(boxNumber)+"()";
                ofs << m_sep;
                ofs << "[" + std::to_string(7 + nDataFields*boxNumber) + "]";
                ofs << "j_low_box_"+std::to_string(boxNumber)+"()";
                ofs << m_sep;
                ofs << "[" + std::to_string(8 + nDataFields*boxNumber) + "]";
                ofs << "k_low_box_"+std::to_string(boxNumber)+"()";
            }
```

Reply:
Nice catch, I moved the separator as suggested.  I think the NaN can be filled the same way.